wp_api/src/comments.rs

```diff
@@ -0,0 +1,497 @@
+    #[case(generate!(CommentListParams, (exclude, vec![CommentId(1), CommentId(2)])), "exclude=1%2C2")]
+    #[case(generate!(CommentListParams, (include, vec![CommentId(1), CommentId(2)])), "include=1%2C2")]
+    #[case(generate!(CommentListParams, (offset, Some(2))), "offset=2")]
+    #[case(generate!(CommentListParams, (order, Some(WpApiParamOrder::Asc))), "order=asc")]
+    #[case(generate!(CommentListParams, (order, Some(WpApiParamOrder::Desc))), "order=desc")]
+    #[case(generate!(CommentListParams, (orderby, Some(WpApiParamCommentsOrderBy::Date))), "orderby=date")]
+    #[case(generate!(CommentListParams, (orderby, Some(WpApiParamCommentsOrderBy::DateGmt))), "orderby=date_gmt")]
+    #[case(generate!(CommentListParams, (orderby, Some(WpApiParamCommentsOrderBy::Id))), "orderby=id")]
+    #[case(generate!(CommentListParams, (orderby, Some(WpApiParamCommentsOrderBy::Include))), "orderby=include")]
+    #[case(generate!(CommentListParams, (orderby, Some(WpApiParamCommentsOrderBy::Post))), "orderby=post")]
+    #[case(generate!(CommentListParams, (orderby, Some(WpApiParamCommentsOrderBy::Parent))), "orderby=parent")]
+    #[case(generate!(CommentListParams, (orderby, Some(WpApiParamCommentsOrderBy::Type))), "orderby=type")]
+    #[case(generate!(CommentListParams, (parent, vec![CommentId(44444), CommentId(44445)])), "parent=44444%2C44445")]
+    #[case(generate!(CommentListParams, (parent_exclude, vec![CommentId(55555), CommentId(55556)])), "parent_exclude=55555%2C55556")]
+    #[case(generate!(CommentListParams, (post, vec![PostId(66666), PostId(66667)])), "post=66666%2C66667")]
+    #[case(generate!(CommentListParams, (status, Some(CommentStatus::Hold))), "status=hold")]
+    #[case(generate!(CommentListParams, (status, Some(CommentStatus::Approve))), "status=approve")]
+    #[case(generate!(CommentListParams, (status, Some(CommentStatus::Spam))), "status=spam")]
+    #[case(generate!(CommentListParams, (status, Some(CommentStatus::Trash))), "status=trash")]
+    #[case(generate!(CommentListParams, (status, Some(CommentStatus::Custom("foo".to_string())))), "status=foo")]
+    #[case(generate!(CommentListParams, (comment_type, Some(CommentType::Comment))), "type=comment")]
+    #[case(generate!(CommentListParams, (comment_type, Some(CommentType::Pingback))), "type=pingback")]
+    #[case(generate!(CommentListParams, (comment_type, Some(CommentType::Trackback))), "type=trackback")]
+    #[case(generate!(CommentListParams, (comment_type, Some(CommentType::Custom("foo".to_string())))), "type=foo")]
+    #[case(generate!(CommentListParams, (password, Some("foo".to_string()))), "password=foo")]
+    #[case(CommentListParams {
+            page: Some(11),
+            per_page: Some(22),
+            search: Some("s_q".to_string()),
+            after: Some("d_a".to_string()),
+            author: vec![UserId(111), UserId(112)],
+            author_exclude: vec![UserId(211), UserId(212)],
+            author_email: Some("a_email".to_string()),
```

Comment:
Question: how should an email like `foo+bar@example.com` be encoded as a URL query? As it is, or `foo%2B@example.com`? Does WordPress accept both?

Similarly, if the user types "Comments about + sign" to search, what's the expected content of the `search` query?

Reply:
> Question: how should an email like foo+bar@example.com be encoded as a URL query? As it is, or foo%2B@example.com?

It'll be encoded as `foo%2Bbar%40example.com`

> Does WordPress accept both?

Although I don't know the answer to this question, I think our current implementation works as expected. For example, if I update the [`email` field used in create user integration test](https://github.com/Automattic/wordpress-rs/blob/2f2cf1d22c196623ec5fcfa8db6c608134ada9a0/wp_api_integration_tests/tests/test_users_mut.rs#L14) to `t_email+foo@example.com`, the created user has the correct email address:

![Screenshot 2024-12-09 at 12 09 01](https://github.com/user-attachments/assets/fa97cdd8-cc8d-43d8-a09e-d6662864323e)

> Similarly, if the user types "Comments about + sign" to search, what's the expected content of the search query?

Pretty much every field will get url encoded by [the `url` crate](https://docs.rs/url/latest/url/) which follows the [URL standard.](https://url.spec.whatwg.org/)

I've added a few more tests about these in 6c3d86c61018733c3b7a18993f43afb6438e34db.

Reply:
> Pretty much every field will get url encoded by [the url crate](https://docs.rs/url/latest/url/) which follows the [URL standard.](https://url.spec.whatwg.org/)

The reason I asked was different web apps treat `+` differently and we need to make sure we are encoding URL queries the way WordPress expects. For example, Google encodes whitespaces as '+' ([example](https://www.google.com/search?q=foo+bar)), instead of its percentage encode value '%20', which is against "the standard". That means a URL like `https://www.google.com/search?q=foo+bar%40email.com` makes Google think the search is `foo bar@email.com`

> The query percent-encode set is the [C0 control percent-encode set](https://url.spec.whatwg.org/#c0-control-percent-encode-set) and U+0020 SPACE, U+0022 ("), U+0023 (#), U+003C (<), and U+003E (>).

Reply:
As shown in the screenshot above, I think we are doing it correctly.